REACT_STYLEGUIDE.MD

```diff
@@ -45,14 +44,31 @@ Bad:
     };
   },
   onClick() {
-    this.setState({ clickCount: this.state.clickCount++ });
+    this.setState(prevState => ({
+      clickCount: prevState.clickCount + 1
+    }));
   },
-  render({ onClick }) {
+  render() {
     return <button className="kuiButton" onClick={this.onClick}/>
   }
 });
 ```
 
+### When a state change only involves the previous state, pass setState a function instead of an object.
```

Comment:
The [official justification](https://facebook.github.io/react/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous) for the second form is that it allows updates that use both `this.state` and `this.props` to be consistent even with async state and props updates.

An added advantage I see is the removal of a `this.state` reference (with `this` being somewhat "dangerous" when not scoped properly in nested functions) even if the props are not used.